akka-remote/src/main/scala/akka/remote/artery/SSLEngineProvider.scala

```diff
@@ -0,0 +1,222 @@
+        val factory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm)
+        factory.init(loadKeystore(SSLKeyStore, SSLKeyStorePassword), SSLKeyPassword.toCharArray)
+        factory.getKeyManagers
+      }
+      val trustManagers = {
+        val trustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm)
+        trustManagerFactory.init(loadKeystore(SSLTrustStore, SSLTrustStorePassword))
+        trustManagerFactory.getTrustManagers
+      }
+      val rng = createSecureRandom()
+
+      val ctx = SSLContext.getInstance(SSLProtocol)
+      ctx.init(keyManagers, trustManagers, rng)
+      ctx
+    } catch {
+      case e: FileNotFoundException ⇒
+        throw new SslTransportException("Server SSL connection could not be established because key store could not be loaded", e)
+      case e: IOException ⇒
+        throw new SslTransportException("Server SSL connection could not be established because: " + e.getMessage, e)
+      case e: GeneralSecurityException ⇒
+        throw new SslTransportException("Server SSL connection could not be established because SSL context could not be constructed", e)
+    }
+  }
+
+  def createSecureRandom(): SecureRandom = {
+    val rng = SSLRandomNumberGenerator match {
+      case r @ ("AES128CounterSecureRNG" | "AES256CounterSecureRNG") ⇒
+        log.debug("SSL random number generator set to: {}", r)
+        SecureRandom.getInstance(r, AkkaProvider)
+      case s @ ("SHA1PRNG" | "NativePRNG") ⇒
+        log.debug("SSL random number generator set to: {}", s)
+        // SHA1PRNG needs /dev/urandom to be the source on Linux to prevent problems with /dev/random blocking
+        // However, this also makes the seed source insecure as the seed is reused to avoid blocking (not a problem on FreeBSD).
+        SecureRandom.getInstance(s)
+
+      case "" ⇒
+        log.debug("SSLRandomNumberGenerator not specified, falling back to SecureRandom")
+        new SecureRandom
+
+      case unknown ⇒
+        log.warning(LogMarker.Security, "Unknown SSLRandomNumberGenerator [{}] falling back to SecureRandom", unknown)
+        new SecureRandom
+    }
+    rng.nextInt() // prevent stall on first access
+    rng
+  }
+
+  private def createSSLEngine(role: TLSRole): SSLEngine = {
+    createSSLEngine(getOrCreateContext(), sslConfig = None, role)
+  }
+
+  override def createServerSSLEngine(): SSLEngine =
+    createSSLEngine(akka.stream.Server)
+
+  override def createClientSSLEngine(): SSLEngine =
+    createSSLEngine(akka.stream.Client)
+
+  private def createSSLEngine(
+    sslContext: SSLContext,
+    sslConfig:  Option[AkkaSSLConfig],
+    role:       TLSRole,
+    closing:    TLSClosing            = IgnoreComplete,
+    hostInfo:   Option[(String, Int)] = None): SSLEngine = {
+
+    /* FIXME this is what NettySSLSupport does
+    val sslEngine = settings.getOrCreateContext(log).createSSLEngine // TODO: pass host information to enable host verification
```

Comment:
We had people asking for host name verification. That would however require one certificate per host with certificates that were issued especially for each host with IP/host-names inside the certificate. We could check what happens when that name information is missing in certificates but I guess it will fail then.

What we might do in the future is adding another flag in the config that would enable host-name verification. In that case, we would pass in the host/port here.